src/pmhn/_trees/_tree_utils.py

```diff
@@ -0,0 +1,189 @@
+from anytree import Node, RenderTree, LevelOrderGroupIter
+from itertools import combinations, product
+from typing import Optional
+
+
+def all_combinations_of_elements(*lists):
+    """
+    Takes a variable number of lists as input and returns a list
+    containing all possible combination of the input lists. In our
+    use case: It takes a list of lists of subtrees (for each child
+    one list of subtrees) as input where a subtree itself is a list
+    of nodes and outputs all combinations of subtrees.
+
+    Args:
+        *lists: any number of lists
+
+    Returns:
+           all combinations of the input lists
+    """
+    n = len(lists)
+    all_combinations = []
+
+    for r in range(1, n + 1):
+        for list_combination in combinations(lists, r):
+            for element_combination in product(*list_combination):
+                all_combinations.append(list(element_combination))
+
+    return all_combinations
```

Comment:
I'm not sure where this function is used, but perhaps you can construct a [lazy](https://en.wikipedia.org/wiki/Lazy_evaluation) generator rather than an explicit list.

In Python there's [yield](https://realpython.com/introduction-to-python-generators/) keyword for this.